tests/test_gpu_examples.py

```diff
@@ -4136,75 +4139,218 @@ def test_prefix_tuning_multiple_devices_encoder_decoder_model(self):
 )
 @pytest.mark.single_gpu_tests
 class TestHotSwapping:
-    @pytest.mark.parametrize("ranks", ["7,13", "13,7"])  # the ranks of the 2 LoRAs as str
-    def test_hotswapping_compiled_model_does_not_trigger_recompilation(self, ranks):
-        env = os.environ.copy()
-        env["TORCH_LOGS"] = "guards,recompiles"
-        here = os.path.dirname(__file__)
-        file_name = os.path.join(here, "run_compiled_model_hotswap.py")
-
-        process = subprocess.Popen(
-            [sys.executable, file_name, "1", ranks], env=env, stdout=subprocess.PIPE, stderr=subprocess.PIPE
-        )
+    """
+    Test hotswapping on compiled models.
 
-        # Communicate will read the output and error streams, preventing deadlock
-        _, stderr = process.communicate()
-        exit_code = process.returncode
+    This test suite is only run on GPU as it is quite slow.
+    """
 
-        # sanity check:
-        assert exit_code == 0
+    torch_device = infer_device()
+    used_model_ids = set()
+
+    def _check_model_id(self, model_id):
+        # Unfortunately, when using the same model_id twice in the same process, even if the LoRA params differ, we get
+        # compilation errors from torch. Therefore, we check that the same ID is not used twice.
+        # Other things I tried that didn't help:
+        # - Setting torch._inductor.config.force_disable_caches = True
+        # - Setting TORCHINDUCTOR_FORCE_DISABLE_CACHES=1
+        # - Setting a separate TORCHINDUCTOR_CACHE_DIR for each test
+        if model_id in self.used_model_ids:
+            raise ValueError("Don't reuse the same model_id twice, as this leads to torch.compile using cache")
+        self.used_model_ids.add(model_id)
+
+    @contextmanager
+    def raise_error_on_recompile(self):
+        """Raise an error when torch recompiles in the context.
+
+        Raises a torch._dynamo.exc.RecompileError error.
+        """
+        prev_value = torch._dynamo.config.error_on_recompile
+        torch._dynamo.config.error_on_recompile = True
+        try:
+            yield
+        finally:
+            torch._dynamo.config.error_on_recompile = prev_value
+
+    #######
+    # LLM #
+    #######
+
+    def check_hotswap(self, do_hotswap, model_id, ranks, alpha_scalings):
```

Comment:
Maybe introduce a context manager that calls [`torch._dynamo.reset()`](https://github.com/pytorch/pytorch/blob/fd0cd6a08f706b7bb1dedb296217b6441e4fb9ff/torch/_dynamo/__init__.py#L85) to reset the caches will work around the need for different models?